src/coreclr/gc/gc.cpp

```diff
@@ -10526,7 +10568,8 @@ bool gc_heap::initial_make_uoh_regions (int gen, gc_heap* hp)
     uint8_t* region_start;
     uint8_t* region_end;
 
-    if (!global_region_allocator.allocate_large_region (&region_start, &region_end))
+    // REGIONS TODO: allocate POH regions on the right
+    if (!global_region_allocator.allocate_large_region (&region_start, &region_end, 1))
```

Comment:
what sort of testing has been done on this? have you tried Server GC?

Reply:
This change involves mostly just the `region_allocator`, and indirectly testing it through managed code is too inefficient, so I just wrote a few unit tests. In particular, these are the lines that I placed right after the `global_region_allocator` is constructed and execution of them is traced:

```c++
uint8_t* end;
uint8_t *v1, *v2, *v3, *v4, *v5, *v6;
global_region_allocator.allocate_large_region(&v1, &end, allocate_forward);
global_region_allocator.allocate_large_region(&v2, &end, allocate_forward);
global_region_allocator.allocate_large_region(&v3, &end, allocate_forward);
global_region_allocator.allocate_large_region(&v4, &end, allocate_forward);
global_region_allocator.allocate_large_region(&v5, &end, allocate_forward);
global_region_allocator.allocate_large_region(&v6, &end, allocate_forward);
global_region_allocator.delete_region(v2);
global_region_allocator.delete_region(v4);
global_region_allocator.delete_region(v3);
global_region_allocator.delete_region(v5);
global_region_allocator.delete_region(v1);
global_region_allocator.delete_region(v6);
```
The test case is designed to exercise through all sorts of free block coalescing behavior. Here is the annotated trace obtained from the execution.

```
[UH]-----printing----before alloc
[UH]-----end printing----[65535 total, left used 0, right used 0]

[UH]-----printing----alloc: found at the end
[UH][B: 8]0->8
[UH]-----end printing----[65535 total, left used 8, right used 0]

[UH]-----printing----alloc: found at the end
[UH][B: 8]0->8
[UH][B: 8]8->16
[UH]-----end printing----[65535 total, left used 16, right used 0]

[UH]-----printing----alloc: found at the end
[UH][B: 8]0->8
[UH][B: 8]8->16
[UH][B: 8]16->24
[UH]-----end printing----[65535 total, left used 24, right used 0]

[UH]-----printing----alloc: found at the end
[UH][B: 8]0->8
[UH][B: 8]8->16
[UH][B: 8]16->24
[UH][B: 8]24->32
[UH]-----end printing----[65535 total, left used 32, right used 0]

[UH]-----printing----alloc: found at the end
[UH][B: 8]0->8
[UH][B: 8]8->16
[UH][B: 8]16->24
[UH][B: 8]24->32
[UH][B: 8]32->40
[UH]-----end printing----[65535 total, left used 40, right used 0]

[UH]-----printing----alloc: found at the end
[UH][B: 8]0->8
[UH][B: 8]8->16
[UH][B: 8]16->24
[UH][B: 8]24->32
[UH][B: 8]32->40
[UH][B: 8]40->48
[UH]-----end printing----[65535 total, left used 48, right used 0]

[UH]-----printing----after delete
[UH][B: 8]0->8
[UH][F: 8]8->16
[UH][B: 8]16->24
[UH][B: 8]24->32
[UH][B: 8]32->40
[UH][B: 8]40->48
[UH]-----end printing----[65535 total, left used 48, right used 0]

[UH]-----printing----after delete
[UH][B: 8]0->8
[UH][F: 8]8->16
[UH][B: 8]16->24
[UH][F: 8]24->32
[UH][B: 8]32->40
[UH][B: 8]40->48
[UH]-----end printing----[65535 total, left used 48, right used 0]

[UH]-----printing----after delete [This is where we are coalescing both the before and the after blocks]
[UH][B: 8]0->8
[UH][F: 24]8->32
[UH][B: 8]32->40
[UH][B: 8]40->48
[UH]-----end printing----[65535 total, left used 48, right used 0]

[UH]-----printing----after delete [This is where we are coalescing the before block only]
[UH][B: 8]0->8
[UH][F: 32]8->40
[UH][B: 8]40->48
[UH]-----end printing----[65535 total, left used 48, right used 0]

[UH]-----printing----after delete [This is where we are coalescing the next block only]
[UH][F: 40]0->40
[UH][B: 8]40->48
[UH]-----end printing----[65535 total, left used 48, right used 0]

[UH]-----printing----after delete [This is where we are reclaiming all space]
[UH]-----end printing----[65535 total, left used 0, right used 0]
```

I also tested the same scenario on the right-hand side, and here is what we have:

```
[UH]-----printing----before alloc
[UH]-----end printing----[65535 total, left used 0, right used 0]

[UH]-----printing----alloc: found at the end
[UH][B: 8]65527->65535
[UH]-----end printing----[65535 total, left used 0, right used 8]

[UH]-----printing----alloc: found at the end
[UH][B: 8]65519->65527
[UH][B: 8]65527->65535
[UH]-----end printing----[65535 total, left used 0, right used 16]

[UH]-----printing----alloc: found at the end
[UH][B: 8]65511->65519
[UH][B: 8]65519->65527
[UH][B: 8]65527->65535
[UH]-----end printing----[65535 total, left used 0, right used 24]

[UH]-----printing----alloc: found at the end
[UH][B: 8]65503->65511
[UH][B: 8]65511->65519
[UH][B: 8]65519->65527
[UH][B: 8]65527->65535
[UH]-----end printing----[65535 total, left used 0, right used 32]

[UH]-----printing----alloc: found at the end
[UH][B: 8]65495->65503
[UH][B: 8]65503->65511
[UH][B: 8]65511->65519
[UH][B: 8]65519->65527
[UH][B: 8]65527->65535
[UH]-----end printing----[65535 total, left used 0, right used 40]

[UH]-----printing----alloc: found at the end
[UH][B: 8]65487->65495
[UH][B: 8]65495->65503
[UH][B: 8]65503->65511
[UH][B: 8]65511->65519
[UH][B: 8]65519->65527
[UH][B: 8]65527->65535
[UH]-----end printing----[65535 total, left used 0, right used 48]

[UH]-----printing----after delete
[UH][B: 8]65487->65495
[UH][B: 8]65495->65503
[UH][B: 8]65503->65511
[UH][B: 8]65511->65519
[UH][F: 8]65519->65527
[UH][B: 8]65527->65535
[UH]-----end printing----[65535 total, left used 0, right used 48]

[UH]-----printing----after delete
[UH][B: 8]65487->65495
[UH][B: 8]65495->65503
[UH][F: 8]65503->65511
[UH][B: 8]65511->65519
[UH][F: 8]65519->65527
[UH][B: 8]65527->65535
[UH]-----end printing----[65535 total, left used 0, right used 48]

[UH]-----printing----after delete [This is where we are coalescing both the before and the after blocks]
[UH][B: 8]65487->65495
[UH][B: 8]65495->65503
[UH][F: 24]65503->65527
[UH][B: 8]65527->65535
[UH]-----end printing----[65535 total, left used 0, right used 48]

[UH]-----printing----after delete [This is where we are coalescing the next block only]
[UH][B: 8]65487->65495
[UH][F: 32]65495->65527
[UH][B: 8]65527->65535
[UH]-----end printing----[65535 total, left used 0, right used 48]

[UH]-----printing----after delete [This is where we are coalescing the before block only]
[UH][B: 8]65487->65495
[UH][F: 40]65495->65535
[UH]-----end printing----[65535 total, left used 0, right used 48]

[UH]-----printing----after delete [This is where we are reclaiming all space]
[UH]-----end printing----[65535 total, left used 0, right used 0]
```